### PR TITLE
Don't adjust flow control windows for streams that don't exist.

### PR DESCRIPTION
**Pull request by Lukasa**
See https://twistedmatrix.com/trac/ticket/8558 for more.

### CODE REVIEW

File: twisted/web/_http2.py [hunk 1]
```diff
@@ -579,6 +579,9 @@ def _handleWindowUpdate(self, event):
         if streamID:
             # Update applies only to a specific stream. If we don't have the
             # stream, that's ok: just ignore it.
+            if streamID not in self.streams:
+                return
```
Comment (adiroiban) [inline twisted/web/_http2.py:584]: maybe would help to get a comment here.. like

```
# Stream no longer registered for flow control.
```

I better way would be to have

```
if self.isStreamRegistered(streamID):
   return

or 

if self.isStreamActive(streamID):
   return
```

please note that I have no idea about the internal architecture and design of HTTP2 ...so my suggestions might be wrong

the important think is that the current `if streamID not in self.streams:` is very generic and I don't think that it expresses what is going on there

File: twisted/web/_http2.py [hunk 2]
```diff
@@ -577,8 +577,13 @@ def _handleWindowUpdate(self, event):
         streamID = event.stream_id
 
         if streamID:
-            # Update applies only to a specific stream. If we don't have the
-            # stream, that's ok: just ignore it.
+            # Update applies only to a specific stream. We may have already
```
Comment (adiroiban) [inline twisted/web/_http2.py]: I think that the `Update applies only to a specific stream.` part is not adding any new information.

File: twisted/web/_http2.py [hunk 3]
```diff
@@ -577,8 +577,13 @@ def _handleWindowUpdate(self, event):
...
+            # cleaned up our stream state, making this a late WINDOW_UPDATE
+            # frame. That's fine, it's not a problem: the update is unnecessary
+            # but benign. We'll just ignore it.
+            if not self._streamIsActive(streamID):
+                return
```
Comment (adiroiban) [inline twisted/web/_http2.py:584]: I think that the 'We might have a late WINDOWS_UPDATE` comment should be in this block.

For me `if not self._streamIsActive(streamID):` speak for itself and we no longer need the verbose comment above

`Update applies only to a specific stream. We may have already cleaned up our stream state`

Comment (Lukasa) [inline twisted/web/_http2.py:584]: I've further clarified the comment.

**Pull request merged**